compilation error magnum and fcl

Hello,
I have the following compilation error : 
```
[37/51] Compiling src/robot_dart/gui/magnum/resources/resources.conf
In file included from ../../../rf_ws/devel/include/dart/collision/fcl/BackwardCompatibility.hpp:87:0,
                 from ../../../rf_ws/devel/include/dart/collision/fcl/FCLTypes.hpp:37,
                 from ../../../rf_ws/devel/include/dart/collision/fcl/FCLCollisionDetector.hpp:38,
                 from ../src/examples/meshes.cpp:7:
/usr/include/fcl/octree.h: In member function ‘fcl::OcTree::OcTreeNode* fcl::OcTree::getNodeChild(fcl::OcTree::OcTreeNode*, unsigned int)’:
/usr/include/fcl/octree.h:198:18: error: ‘const class octomap::OcTree’ has no member named ‘getNodeChild’
     return tree->getNodeChild(node, childIdx);
                  ^
/usr/include/fcl/octree.h: In member function ‘const OcTreeNode* fcl::OcTree::getNodeChild(const OcTreeNode*, unsigned int) const’:
/usr/include/fcl/octree.h:208:18: error: ‘const class octomap::OcTree’ has no member named ‘getNodeChild’
     return tree->getNodeChild(node, childIdx);
                  ^
/usr/include/fcl/octree.h: In member function ‘bool fcl::OcTree::nodeChildExists(const OcTreeNode*, unsigned int) const’:
/usr/include/fcl/octree.h:218:18: error: ‘const class octomap::OcTree’ has no member named ‘nodeChildExists’
     return tree->nodeChildExists(node, childIdx);
                  ^
/usr/include/fcl/octree.h: In member function ‘bool fcl::OcTree::nodeHasChildren(const OcTreeNode*) const’:
/usr/include/fcl/octree.h:228:18: error: ‘const class octomap::OcTree’ has no member named ‘nodeHasChildren’
     return tree->nodeHasChildren(node);
                  ^

In file included from ../../../rf_ws/devel/include/dart/collision/fcl/BackwardCompatibility.hpp:87:0,
                 from ../../../rf_ws/devel/include/dart/collision/fcl/FCLTypes.hpp:37,
                 from ../../../rf_ws/devel/include/dart/collision/fcl/FCLCollisionDetector.hpp:38,
                 from ../src/examples/meshes.cpp:7:
/usr/include/fcl/octree.h: In member function ‘fcl::OcTree::OcTreeNode* fcl::OcTree::getNodeChild(fcl::OcTree::OcTreeNode*, unsigned int)’:
/usr/include/fcl/octree.h:198:18: error: ‘const class octomap::OcTree’ has no member named ‘getNodeChild’
     return tree->getNodeChild(node, childIdx);
                  ^
/usr/include/fcl/octree.h: In member function ‘const OcTreeNode* fcl::OcTree::getNodeChild(const OcTreeNode*, unsigned int) const’:
/usr/include/fcl/octree.h:208:18: error: ‘const class octomap::OcTree’ has no member named ‘getNodeChild’
     return tree->getNodeChild(node, childIdx);
                  ^
/usr/include/fcl/octree.h: In member function ‘bool fcl::OcTree::nodeChildExists(const OcTreeNode*, unsigned int) const’:
/usr/include/fcl/octree.h:218:18: error: ‘const class octomap::OcTree’ has no member named ‘nodeChildExists’
     return tree->nodeChildExists(node, childIdx);
                  ^
/usr/include/fcl/octree.h: In member function ‘bool fcl::OcTree::nodeHasChildren(const OcTreeNode*) const’:
/usr/include/fcl/octree.h:228:18: error: ‘const class octomap::OcTree’ has no member named ‘nodeHasChildren’
     return tree->nodeHasChildren(node);


```
It is certainly due to a fcl versioning issue

On my computer I have
```
libfcl-0.5-dev/now 0.5.0-3osrf1~xenial3 amd64 [installed,local]
  Flexible Collision Library - development files

liboctomap-dev/xenial,now 1.6.8+dfsg-2.1 amd64 [installed]
  Octomap library development files

ros-kinetic-octomap/now 1.8.1-0xenial-20180809-134711-0800 amd64 [installed,local]
  The OctoMap library implements a 3D occupancy grid mapping approach, providing data structures and mapping algorithms in C++.

ros-kinetic-octomap-msgs/now 0.3.3-0xenial-20181107-010952-0800 amd64 [installed,local]
  This package provides messages and serializations / conversion for the OctoMap library.

ros-kinetic-octomap-ros/now 0.4.0-0xenial-20181117-015315-0800 amd64 [installed,local]
  octomap_ros provides conversion functions between ROS and OctoMap's native types.
```

I cannot do as described in the robot_dart installation process : 
``` sudo apt-get install libfcl-dev```
 Because it would uninstall a lot of ros packages

Ok so actually the problem is with octomap
I can install dart using libfcl-0.5-dev and the embedded ros-kinetic liboctomap.so.1.8.1

When it is like this I now get this error with robot_dart 
```
[30/51] Compiling src/examples/cameras.cpp
In file included from ../../dart_install/include/dart/collision/fcl/BackwardCompatibility.hpp:87:0,
                 from ../../dart_install/include/dart/collision/fcl/FCLTypes.hpp:37,
                 from ../../dart_install/include/dart/collision/fcl/FCLCollisionDetector.hpp:38,
                 from ../src/examples/meshes.cpp:7:
/usr/include/fcl/octree.h:50:29: fatal error: octomap/octomap.h: No such file or directory
compilation terminated.
```

If I install 
```
sudo apt-get install liboctomap-dev
```
It will install the liboctomap1.6.8 version 
If I use this and not the liboctomap1.8.1 I will have the same following error when installing dart 
```
In file included from ../../../rf_ws/devel/include/dart/collision/fcl/BackwardCompatibility.hpp:87:0,
                 from ../../../rf_ws/devel/include/dart/collision/fcl/FCLTypes.hpp:37,
                 from ../../../rf_ws/devel/include/dart/collision/fcl/FCLCollisionDetector.hpp:38,
                 from ../src/examples/meshes.cpp:7:
/usr/include/fcl/octree.h: In member function ‘fcl::OcTree::OcTreeNode* fcl::OcTree::getNodeChild(fcl::OcTree::OcTreeNode*, unsigned int)’:
/usr/include/fcl/octree.h:198:18: error: ‘const class octomap::OcTree’ has no member named ‘getNodeChild’
     return tree->getNodeChild(node, childIdx);
```

So in the end, the problem is that robot_dart cannot find 
/opt/ros/kinetic/include/octomap/octomap.h

--- a/waf_tools/dart.py
+++ b/waf_tools/dart.py
@@ -82,11 +82,36 @@ def check_dart(conf, *k, **kw):
     except:
         ode_found = False
 
+    # DART has some optional Octomap dependency
+    octomap_check = ['/usr/local/include', '/usr/include']
+    octomap_libs = ['/usr/local/lib', '/usr/local/lib64', '/usr/lib', '/usr/lib64', '/usr/lib/x86_64-linux-gnu/']
+    if 'ROS_DISTRO' in os.environ:
+        octomap_check.append('/opt/ros/' + os.environ['ROS_DISTRO'] + '/include')
+        octomap_libs.append('/opt/ros/' + os.environ['ROS_DISTRO'] + '/lib')
+    octomap_include = []
+    octomap_lib = []
+    octomap_found = False
+    try:
+        octomap_include = [get_directory('octomap/octomap.h', octomap_check)]
+        octomap_lib = [get_directory('liboctomap.' + suffix, octomap_libs)]
+        octomap_found = True
+    except:
+        octomap_found = False
+
     dart_load_prefix = 'utils'
     dart_include = []
     dart_major = -1
     dart_minor = -1
     dart_patch = -1
+    # DART Configurations
+    dart_have_nlopt = False
+    dart_have_ipopt = False
+    dart_have_pagmo = False
+    dart_have_snopt = False
+    dart_have_bullet = False
+    dart_have_ode = False
+    dart_have_flann = False
+    dart_have_octomap = False
 
     try:
         conf.start_msg('Checking for DART includes (including io/urdf)')
@@ -105,6 +130,40 @@ def check_dart(conf, *k, **kw):
                 dart_patch = int(line.split(' ')[-1].strip())
             if dart_major > 0 and dart_minor > 0  and dart_patch > 0:
                 break
+
+            nlopt = line.find('#define HAVE_NLOPT')
+            ipopt = line.find('#define HAVE_IPOPT')
+            pagmo = line.find('#define HAVE_PAGMO')
+            snopt = line.find('#define HAVE_SNOPT')
+            bullet = line.find('#define HAVE_BULLET')
+            ode = line.find('#define HAVE_ODE')
+            flann = line.find('#define HAVE_FLANN')
+            octomap = line.find('#define HAVE_OCTOMAP')
+
+            if nlopt > -1:
+                boolean = int(line.split(' ')[-1].strip())
+                dart_have_nlopt = (boolean == 1)
+            if ipopt > -1:
+                boolean = int(line.split(' ')[-1].strip())
+                dart_have_ipopt = (boolean == 1)
+            if pagmo > -1:
+                boolean = int(line.split(' ')[-1].strip())
+                dart_have_pagmo = (boolean == 1)
+            if snopt > -1:
+                boolean = int(line.split(' ')[-1].strip())
+                dart_have_snopt = (boolean == 1)
+            if bullet > -1:
+                boolean = int(line.split(' ')[-1].strip())
+                dart_have_bullet = (boolean == 1)
+            if ode > -1:
+                boolean = int(line.split(' ')[-1].strip())
+                dart_have_ode = (boolean == 1)
+            if flann > -1:
+                boolean = int(line.split(' ')[-1].strip())
+                dart_have_flann = (boolean == 1)
+            if octomap > -1:
+                boolean = int(line.split(' ')[-1].strip())
+                dart_have_octomap = (boolean == 1)
 
         if dart_major < 6:
             raise Exception('We need DART version at least 6.0.0')
@@ -145,34 +204,49 @@ def check_dart(conf, *k, **kw):
         else:
             conf.end_msg('Not found - Your programs may not compile', 'RED')
 
-        conf.start_msg('DART: Checking for Bullet Collision libs')
-        if bullet_found:
-            try:
-                bullet_lib.append(get_directory('libdart-collision-bullet.'+suffix, libs_check))
-                conf.env.INCLUDES_DART = conf.env.INCLUDES_DART + bullet_include
-                conf.env.LIBPATH_DART =  conf.env.LIBPATH_DART + bullet_lib
-                conf.env.LIB_DART.append('LinearMath')
-                conf.env.LIB_DART.append('BulletCollision')
-                conf.env.LIB_DART.append('dart-collision-bullet')
-                conf.end_msg('libs: ' + str(conf.env.LIB_DART[-3:]) + ', bullet: ' + str(bullet_include[0]))
-            except:
+        if dart_have_bullet:
+            conf.start_msg('DART: Checking for Bullet Collision libs')
+            if bullet_found:
+                try:
+                    bullet_lib.append(get_directory('libdart-collision-bullet.'+suffix, libs_check))
+                    conf.env.INCLUDES_DART = conf.env.INCLUDES_DART + bullet_include
+                    conf.env.LIBPATH_DART =  conf.env.LIBPATH_DART + bullet_lib
+                    conf.env.LIB_DART.append('LinearMath')
+                    conf.env.LIB_DART.append('BulletCollision')
+                    conf.env.LIB_DART.append('dart-collision-bullet')
+                    conf.end_msg('libs: ' + str(conf.env.LIB_DART[-3:]) + ', bullet: ' + str(bullet_include[0]))
+                except:
+                    conf.end_msg('Not found', 'RED')
+            else:
                 conf.end_msg('Not found', 'RED')
-        else:
-            conf.end_msg('Not found', 'RED')
 
-        conf.start_msg('DART: Checking for Ode Collision libs')
-        if ode_found:
-            try:
-                ode_lib.append(get_directory('libdart-collision-ode.'+suffix, libs_check))
-                conf.env.INCLUDES_DART = conf.env.INCLUDES_DART + ode_include
-                conf.env.LIBPATH_DART =  conf.env.LIBPATH_DART + ode_lib
-                conf.env.LIB_DART.append('ode')
-                conf.env.LIB_DART.append('dart-collision-ode')
-                conf.end_msg('libs: ' + str(conf.env.LIB_DART[-2:]) + ', ode: ' + str(ode_include[0]))
-            except:
+        if dart_have_ode:
+            conf.start_msg('DART: Checking for Ode Collision libs')
+            if ode_found:
+                try:
+                    ode_lib.append(get_directory('libdart-collision-ode.'+suffix, libs_check))
+                    conf.env.INCLUDES_DART = conf.env.INCLUDES_DART + ode_include
+                    conf.env.LIBPATH_DART =  conf.env.LIBPATH_DART + ode_lib
+                    conf.env.LIB_DART.append('ode')
+                    conf.env.LIB_DART.append('dart-collision-ode')
+                    conf.end_msg('libs: ' + str(conf.env.LIB_DART[-2:]) + ', ode: ' + str(ode_include[0]))
+                except:
+                    conf.end_msg('Not found', 'RED')
+            else:
                 conf.end_msg('Not found', 'RED')
-        else:
-            conf.end_msg('Not found', 'RED')
+
+        if dart_have_octomap:
+            conf.start_msg('DART: Checking for Octomap libs')
+            if octomap_found:
+                try:
+                    conf.env.INCLUDES_DART = conf.env.INCLUDES_DART + octomap_include
+                    conf.env.LIBPATH_DART =  conf.env.LIBPATH_DART + octomap_lib
+                    conf.env.LIB_DART.append('octomap')
+                    conf.end_msg('libs: ' + str(conf.env.LIB_DART[-1:]) + ', octomap: ' + str(octomap_include[0]))
+                except:
+                    conf.end_msg('Not found', 'RED')
+            else:
+                conf.end_msg('Not found', 'RED')
 
         # remove duplicates
         conf.env.INCLUDES_DART = list(set(conf.env.INCLUDES_DART))

